### PR TITLE
Replace notmuch_databse_open with notmuch_database_open_with_config

### LINKED ISSUE #31: Deprecated notmuch database open function
go.notmuch throw this warning during the compilation. Maybe we should update the to the new binding?

```text
# github.com/zenhack/go.notmuch
cgo-gcc-prolog: In function ‘_cgo_c02956f23830_Cfunc_notmuch_database_open’:
cgo-gcc-prolog:347:2: warning: ‘notmuch_database_open’ is deprecated: function deprecated as of libnotmuch 5.4 [-Wdeprecated-declarations]
In file included from /build/go/pkg/mod/github.com/brunnre8/go.notmuch@v0.0.0-20201126061756-caa2daf7093c/db.go:9:
/usr/include/notmuch.h:332:1: note: declared here
  332 | notmuch_database_open (const char *path,
      | ^~~~~~~~~~~~~~~~~~~~~
```

### PR DESCRIPTION
**Pull request by stbenjam**
Fixes #31 

In libnotmuch 5.4, notmuch_database_open is a deprecated alias for
notmuch_database_open_with_config with config="" and error=NULL.

This updates the Open function and adds a wrapper OpenWithConfig.

### CODE REVIEW

File: db.go [hunk 1]
```diff
@@ -60,13 +60,23 @@ func Create(path string) (*DB, error) {
 // Open opens the database at the location path using mode. Caller is responsible
 // for closing the database when done.
 func Open(path string, mode DBMode) (*DB, error) {
+	return OpenWithConfig(path, "", mode)
+}
+
+// OpenWithConfig opens the database at the location path using mode using the
+// configuration in config. Caller is responsible for closing the database when
+// done.
+func OpenWithConfig(path, config string, mode DBMode) (*DB, error) {
```
Comment (brunnre8) [inline db.go]: you want to be able to pass nil here, that signature doesn't work.

Empty string and null don't behave the same 

```
390  * @param[in]  config_path
[ 391](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l391)  * @parblock
[ 392](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l392)  * Path to config file.
[ 393](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l393)  *
[ 394](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l394)  * Config file is key-value, with mandatory sections. See
[ 395](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l395)  * <em>notmuch-config(5)</em> for more information. The key-value pair
[ 396](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l396)  * overrides the corresponding configuration data stored in the
[ 397](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l397)  * database (see <em>notmuch_database_get_config</em>)
[ 398](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l398)  *
[ 399](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l399)  * If <em>config_path</em> is NULL use the path specified
[ 400](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l400)  *
[ 401](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l401)  * - in environment variable <em>NOTMUCH_CONFIG</em>, if non-empty
[ 402](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l402)  *
[ 403](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l403)  * - by  <em>XDG_CONFIG_HOME</em>/notmuch/ where
[ 404](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l404)  *   XDG_CONFIG_HOME defaults to "$HOME/.config".
[ 405](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l405)  *
[ 406](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l406)  * - by $HOME/.notmuch-config
[ 407](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l407)  *
[ 408](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l408)  * If <em>config_path</em> is "" (empty string) then do not
[ 409](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l409)  * open any configuration file.
[ 410](https://git.notmuchmail.org/git?p=notmuch;a=blob;f=lib/notmuch.h;hb=HEAD#l410)  * @endparblock
```

Comment (stbenjam) [reply db.go]: Ah, thanks. That's fairly nuanced.  Similar conditions apply for path.  How about this version?  I wasn't sure how to send a go string pointer to C -- is this correct?

File: db.go [hunk 2]
```diff
@@ -60,13 +60,42 @@ func Create(path string) (*DB, error) {
...
+//  - by $XDG_DATA_HOME/notmuch/$PROFILE
+//
+// If config is nil, it will use the path:
+//	- in environment variable NOTMUCH_CONFIG, if non-empty
+//	- by $XDG_CONFIG_HOME/notmuch/
+//
+// If config is an empty string (""), then it will not open any configuration file.
+//
+// Caller is responsible for closing the database when done.
+func OpenWithConfig(path, config *string, mode DBMode) (*DB, error) {
```
Comment (brunnre8) [inline db.go]: This is still missing the profile, you should expose the same API as the underlying notmuch library I think.

File: db_test.go [hunk 3]
```diff
@@ -58,6 +58,28 @@ func TestOpen(t *testing.T) {
...
+		t.Fatalf("CreateTemp(): unexpected error: %s", err)
+	}
+	defer config.Close()
+	defer os.Remove(config.Name())
+	if _, err := config.WriteString(fmt.Sprintf("[database]\npath=%s\n", dbPath)); err != nil {
+		t.Fatalf("WriteString(%q): unexpected error: %s", config.Name(), err)
+	}
+
+	configPath := config.Name()
+	db, err := OpenWithConfig(nil, &configPath, DBReadOnly)
```
Comment (brunnre8) [inline db_test.go]: You aren't testing most of the combinations.
This probably needs a table driven test so that so see all combinations of some string or nil to ensure that this actually works and that you can't regress

File: db.go [hunk 4]
```diff
@@ -60,13 +60,42 @@ func Create(path string) (*DB, error) {
...
+	if config != nil {
+		cconfig = C.CString(*config)
+		defer C.free(unsafe.Pointer(cconfig))
+	}
 
 	cmode := C.notmuch_database_mode_t(mode)
 	var cdb *C.notmuch_database_t
 	cdbptr := (**C.notmuch_database_t)(&cdb)
-	err := statusErr(C.notmuch_database_open(cpath, cmode, cdbptr))
+	err := statusErr(C.notmuch_database_open_with_config(cpath, cmode, cconfig, nil, cdbptr, nil))
```
Comment (brunnre8) [inline db.go]: You are ignoring the error output var, what's your reasoning there?
While I understand that this works, it would be much nicer to add the error message from notmuch, meant for human consumption, to it and let the user know the details rather than just "doesn't work"

File: db.go (hunk 1 above)
Comment (brunnre8) [reply db.go]: yeah dereferencing a non nil value and shoving it in the Cstring function seems ok.

File: db.go [hunk 5]
```diff
@@ -60,13 +60,42 @@ func Create(path string) (*DB, error) {
 // Open opens the database at the location path using mode. Caller is responsible
 // for closing the database when done.
 func Open(path string, mode DBMode) (*DB, error) {
-	cpath := C.CString(path)
-	defer C.free(unsafe.Pointer(cpath))
+	config := ""
```
Comment (brunnre8) [inline db.go:63]: that's probably the wrong default, I'd pass nil

File: db.go (hunk 2 above)
Comment (brunnre8) [reply db.go:82]: oh wait, if that's what upstream is doing (the C lib) by all means ignore me.

We should do the same

File: db.go (hunk 5 above)
Comment (brunnre8) [reply db.go:63]: oh wait, if that's what upstream is doing (the C lib) by all means ignore me.

We should do the same

**Comment by stbenjam** [comment]
@brunnre8 Thanks for your review + attention to details. I've updated this, I think it keeps the same interface as the C version.

File: db.go [hunk 6]
```diff
@@ -60,14 +60,53 @@ func Create(path string) (*DB, error) {
...
+//  - by $XDG_DATA_HOME/notmuch/$PROFILE, if profile argument is set
+//
+// If config is nil, it will use the path:
+//  - in environment variable NOTMUCH_CONFIG, if non-empty
+//  - in $XDG_CONFIG_HOME/notmuch/
+//
+// If config is an empty string (""), then it will not open any configuration file.
+//
+// Caller is responsible for closing the database when done.
+func OpenWithConfig(path, config, profile *string, mode DBMode) (*DB, error) {
```
Comment (brunnre8) [inline db.go:94]: path is not documented yet

Comment (brunnre8) [reply db.go:94]: ehr, profile

File: db.go [hunk 7]
```diff
@@ -60,14 +60,53 @@ func Create(path string) (*DB, error) {
...
+	defer C.free(unsafe.Pointer(cErrMsg))
 
 	cmode := C.notmuch_database_mode_t(mode)
 	var cdb *C.notmuch_database_t
 	cdbptr := (**C.notmuch_database_t)(&cdb)
-	err := statusErr(C.notmuch_database_open(cpath, cmode, cdbptr))
-	if err != nil {
+	err := statusErr(C.notmuch_database_open_with_config(cpath, cmode, cconfig, cprofile, cdbptr, &cErrMsg))
+	if err != nil || errMsg != "" {
 		return nil, err
```
Comment (brunnre8) [inline db.go:110]: This check should never have an impact, notmuch will return an error code and then the reason in it's err msg out param.
As is, this will not have an effect whatsoever to the caller.

You'd need to change the error constructor to pass a message if you want to bubble this up.

File: db.go (hunk 6 above)
Comment (stbenjam) [reply db.go:94]: Fixed

**Comment by zenhack** [comment]
LGTM.

**Comment by zenhack** [comment]
Merged, thanks for the patch!
